test_plans/atomic_ref.asciidoc

```diff
@@ -0,0 +1,480 @@
+
+= Test plan for atomic_ref
+
+This is a test plan for the APIs described in SYCL 2020 section 4.15.3. Atomic references
+
+== Testing scope
+
+The test plan covers non-atomic behavior, the latter will be covered in a separate test plan
+
+=== Device coverage
+
+All of the tests described below are performed on the default device that is selected on the CTS command line
+
+== Type coverage
+
+All of the tests described below are performed using each of the
+following types as the `ReferencedType`.
+
+In Regular mode:
+
+* `int`
+* `float`
+
+In full conformance mode:
+
+* `int`
+* `unsigned int`
+* `long`
+* `unsigned long`
+* `float`
+
+In addition, if the device has `aspect::atomic64`, the following types are tested:
+
+* `long long`
+* `unsigned long long`
+* `double`
+
+Additionaly run same tests for `ReferencedType*`.
+
+Note: using T as `ReferencedType` or `ReferencedType*`.
+
+== Tests
+
+All non static tests run on following cases:
+
+* single_task with referenced value is accessor element - should be skipped for `address_space::local_space`
+* parallel_for with nd_range parameter and referenced value local_accessor element - should be skipped for `address_space::global_space`
+
+Note: all tests run with different values of `DefaultOrder` and `DefaultScope` template parameters, for unsupported values tests are skiped. To check if the special values of `DefaultOrder` parameters are supported should use `info::device::atomic_memory_order_capabilities`, `memory_order::relaxed` should be supported on all devices. To check if the special values for `DefaultScope` parameter are supported should use `info::device::atomic_memory_scope_capabilities`.
+
+=== SYCL2020 interface
+
+==== Members
+
+For `atomic_ref` with the following `DefaultOrder` template parameter values:
+
+* `memory_order::relaxed`
+* `memory_order::acq_rel`
+* `memory_order::seq_cst`
+
+And the following values as `DefaultScope` template parameter:
+
+* `memory_scope::work_item`
+* `memory_scope::sub_group`
+* `memory_scope::work_group`
+* `memory_scope::device`
+* `memory_scope::system`
+
+And the following values as `Space` template parameter:
+
+* `access::address_space::global_space`
+* `access::address_space::local_space`
+* `access::address_space::generic_space`
+
+Check the following:
+
+* `default_read_order`, `default_write_order` and `default_read_modify_write_order` members are const
+* `required_alignment` memeber is const and is equal to or grater than `alignof(ReferencedType)`
+* `is_always_lock_free` member is const and is equal to `true` of `false`
+* `default_scope` member is const and is equal to `DefaultScope` template parameter was used when object was created
+
+If `T` is `ReferencedType` check that `value_type` is same as `ReferencedType`.
+
+If `T` is `ReferencedType*` check that `value_type` is same as `ReferencedType*`.
+
+If `T` is `Integral` or `Floating` check that `atomic_ref::difference_type` is same
+as `atomic_ref::value_type`.
```

Comment:
If I'm not mistaken, `atomic_ref::difference_type` only exists in partial specialization for pointers

Reply:
The spec also defines it for the specializations for integral types and floating point types.